scrapyd_k8s/tests/integration/test_api.py

```diff
@@ -159,6 +163,16 @@ def test_scenario_cancel_running_finished_ok():
     # wait until the job has stopped
     listjobs_wait(jobid, 'finished')
     jobinfo = assert_listjobs(finished=jobid)
+    start_time, end_time = jobinfo.pop('start_time'), jobinfo.pop('end_time')
+    assert start_time and datetime.strptime(start_time, TIME_FORMAT)
+
+    # Only when a k8s job has succeeded will it have a completion time. Here we cancel
+    # the job and thus we expect the end_time to undefined.
+    if isinstance(launcher, K8s):
+        assert not end_time
+    else:
+        assert end_time and datetime.strptime(end_time, TIME_FORMAT)
```

Comment:
I understand that you are tempted to check the `launcher` here.
This is an integration test and it needs to have a well-defined behaviour. No ifs like this here.
If you want to control the job, then we need to either setup the environment to do this, or find another way. E.g. add a spider setting to the example spider that lets it return with a non-success exit code, or raise an error. Or add something to the scrapyd-k8s to schedule a job without starting it.

Do _not_ test if running on k8s or not - if you _really_ need to distinguish behaviour, make it explicit, e.g. add an evironment variable to the test that exactly tells what you expect in the test (e.g. `TEST_END_TIME_MISSING=1`).

It looks like behaviour on Docker vs. Kubernetes different. Why? I don't think we'd want this.

Reply:
Adding the `end_time` will require an entirely different approach. The behaviour is different when using Docker vs. Kubernetes. Reverted the code for now so we can come up with something else. Not sure yet on how to approach this.

Reply:
Ok, let's discuss it in the issue!
We could also include it just for Kubernetes, but then I'd like to document it, and know what would be needed (and probably it could make sense to explain it in the Docker launcher code).